### ISSUE TITLE
No fbcon video output

### ISSUE THREAD

**Issue by OrdinaryMagician** [mpv-player/mpv#108]
It is impossible to get working video output in the Linux framebuffer console.

After the removal of the fbdev drivers (which worked wonders) I know I should use sdl, but not even that works (not to mention it has to be SDL 2 specifically). All I get when starting up is "[sdl] SDL_Init failed" followed by only audio (this doesn't happen on X, I do get working video there).

Other programs work well with SDL 1.2 on the console, so it must be a SDL 2 thing...

If you could please make SDL 1.2 support on mpv not be limited to just audio output, I would be really thankful.

**Comment by parkerlreed** [comment]
+1 for KMS/DRM support. sdl2 on Arch doesn't ship with directfb enabled and from my experience trying to get it enabled in a PITA.

**Comment by rr-** [comment]
@wm4 I'm interested but have absolutely zero experience. I could probably contribute a thing or two under right guidance but I'm not sure if you have time to babysit people.

**Comment by wm4** [comment]
Sure I have. What are you interested in? There's KMS and "classic" Linux framebuffer. They are both very different. For KMS, DRM could be used to render with OpenGL, so it'd be like a vo_opengl backend (like gl_x11.c, gl_wayland.c etc.).

Classic Linux framebuffer (vo_fb_wip branch) would render directly to a linear framebuffer in GPU memory. It has the disadvantage that it's very slow. For one because it requires software conversion and scaling from YUV (video) to RGB, and also because accessing GPU RAM directly is slow. vo_x11.c or vo_wayland.c could serve as a template for such a VO (they also use software scaling).

I don't know if KMS can give you a linear framebuffer. I suppose it might be possible.

**Comment by rr-** [comment]
Thanks, that sheds some light. I'll setup some virtual machines and try to play with mini-programs to get a hang of what it's like (I mean, talking to KMS and directfb). If I manage to do anything of some significance, I'll try to make a vo out of it.

As far as I understand it, the rendering process works like this:

![diagram](https://cloud.githubusercontent.com/assets/1045476/7045494/f8de7388-ddfe-11e4-9125-0939450108d4.png)

...where `vo` is a proxy between the final frame and physical device. Is this right? 

Also, you mentioned that we need to do software conversion of YUV to RGB. Is this because even if we returned RGB in `query_format()`, the codecs wouldn't understand it directly and still output only YUV?

**Comment by wm4** [comment]
>directfb

This is AFAIK a (really thick) abstraction wrapper around Linux fbdev. It might be valid to use, and MPlayer also had support for it, besides using Linux FB directly.

>As far as I understand it, the rendering process works like this:

Almost, but OSD and video are handled within the VO. The VO cares about putting mp_image (the struct which stores video data) on the screen. The VO also has to take care about scaling the video frame, and rendering OSD and subtitles to the screen. Some VOs, like vo_opengl, use the video APIs to do this efficiently, while other VOs (like vo_wayland and vo_x11) render them directly into the image. The latter also use libswscale (though the mpv wrapper function ``mp_sws_scale()``) for video conversion and scaling.

Basically, having to handle scaling and OSD is just a bother for pure software-based VOs, but necessary for VOs which use graphics APIs like OpenGL.

>Also, you mentioned that we need to do software conversion of YUV to RGB. Is this because even if we returned RGB in query_format(), the codecs wouldn't understand it directly and still output only YUV?

You could return only a RGB format. Then the player would add a video filter which converts the decoder's YUV output to RGB. But since you need to scale the result anyway, and since libswscale combines conversion and scaling, it's probably more efficient to do both in the VO.

The main reason the VO has to care about scaling is because the rest of the player barely cares about the window size. MPlayer used to have something to scale to display size before the VO, but it was kind of terrible. And especially with the swscale wrapper, invoking the scaler within the VO becomes relatively trivial.

**Comment by wm4** [comment]
>Almost, but OSD and video are handled within the VO. 

In vo_x11.c, rendering the OSD and subtitles is done with this call:
```C
osd_draw_on_image(vo->osd, p->osd, mpi ? mpi->pts : 0, 0, &img);
```
Where img is the target mp_image.

One complication is that rendering directly into the framebuffer might be slow (because of uncacheable GPU memory), and you might have to copy the image.

I've also looked at vo_wayland. It actually uses wayland sub-surfaces for OSD overlays, so it's much more complicated.

**Comment by rr-** [comment]
Alright, I learned a few things so far:

- One needs `libdrm` to be able to link against libraries that allow talking to KMS, this is the only dependency. I couldn't find any way to talk to KMS without this library.
- I had to do a weird hack with `PATH` to make gcc discover `/usr/include/libdrm/*.h` included by `/usr/include/xf86drm.h`. In mpv's case this could be figured out by `configure`, or I'm doing something wrong altogether.
- There is linear buffer (called "dumb buffer") support, that is said to work with any driver :+1:
- However, apparently the driver provided by VirtualBox doesn't support direct buffers :100: so I ended up using it on my laptop, and it works perfectly - no need for X whatsoever. The minimum requirement appears to have kernel create a `/dev/dri/card0` device, which we refer to in the code. I'm not entirely sure what conditions need to be met for this to happen, but since I have X installed on my laptop with simple vesa drivers, and VirtualBox told me that it doesn't support direct buffers only after I installed guest additions, I suspect this means user needs to have any kind of graphic driver installed.
- User might have a few graphic cards (`/dev/dri/card0`, `/dev/dri/card`, ...), so it would be reasonable to make path to them configurable. The other configurable option that should be exposed is which monitor to use.
- Regarding performance: I can write pixels directly to this buffer, but it's quite slow. For some reason, writing twice to such buffer causes it to lag terribly, probably because it tries to keep up with each change separately. Fortunately DRM offers double buffering and page flipping (I'm not sure what's the difference between these?).
- Besides these dumb buffers, there are other things. With addition of mesa3d and `libgbm` this could even integrate with OpenGL. I'd, however, avoid this approach altogether, since most of this revolves around code specific to different video card vendors, and that is surely going to cause linking headaches and will unnecessarily bloat the program.

For now I'll try to create a proof of concept slow-as-hell implementation, and then focus on performance issues by integrating double buffering and possibly other stuff I might discover along the road.

**Comment by wm4** [comment]
Sounds good. I'll reply to some details tomorrow.

**Comment by wm4** [comment]
>One needs libdrm to be able to link against libraries that allow talking to KMS, this is the only dependency. I couldn't find any way to talk to KMS without this library.

>There is linear buffer (called "dumb buffer") support, that is said to work with any driver

There's also classic Linux framebuffer, for which ioctls can be used directly. It's potentially more portable (but also has accumulated much more cruft). The difference probably doesn't matter, and it would be preferable to use KMS. But it still could be that there's some hardware that works only with classic FB.

The nvidia binary driver doesn't provide KMS, but on the other hand, it doesn't officially support classic fbdev either.

>I had to do a weird hack with PATH to make gcc discover /usr/include/libdrm/*.h included by /usr/include/xf86drm.h. In mpv's case this could be figured out by configure, or I'm doing something wrong altogether.

This sounds wrong. The libdrm .pc file actually adds this include file, so you shouldn't do this manually. Just leave it to pkg-config. (mpv uses pkg-config for a lot of things.) You can check yourself:
```
$ pkg-config libdrm --cflags
-I/usr/include/libdrm 
```

>Regarding performance: I can write pixels directly to this buffer, but it's quite slow. For some reason, writing twice to such buffer causes it to lag terribly, probably because it tries to keep up with each change separately.

Yes, the framebuffer is probably on GPU memory. It's not cached, and every memory access will trigger a PCI transfer or something (or worse, I don't know about the details). Of course this is prohibitively inefficient, and apparently also the reason why "classic" Linux framebuffer is so slow.

>Fortunately DRM offers double buffering and page flipping (I'm not sure what's the difference between these?).

Uh, no idea. Maybe 2 aspects of the same thing?

>Besides these dumb buffers, there are other things. With addition of mesa3d and libgbm this could even integrate with OpenGL. I'd, however, avoid this approach altogether, since most of this revolves around code specific to different video card vendors, and that is surely going to cause linking headaches and will unnecessarily bloat the program.

This is what wayland compositors have to use, so I didn't think it could be so bad.

**Comment by rr-** [comment]
I just found what's the difference between double buffering and page flipping is: double buffering usually involves drawing to a back buffer and then copying its contents to the display buffer. But even that can introduce tearing due to copying being slow, so a technique named "page flipping" was proposed: after drawing to back buffer, you tell the driver it's now a display buffer, and use the previous display buffer as the new back buffer. There's no copying whatsoever, so both buffer need to reside on the GPU.

And then there's triple buffering, that is supposed to make GPU and CPU cycles completely independent. In my understanding, it assumes the drawing algorithm to be asynchronous.

**Comment by rr-** [comment]
Everything  goes well so far.
For now, I added noise image output as a placeholder.

https://github.com/rr-/mpv/blob/drm/video/out/vo_drm.c#L298-L309

**Comment by wm4** [comment]
Looks generally not bad!

One thing, IMGFMT_BGR24 will probably be very slow. It's a 3 byte format, so it's hard to process it efficiently with SIMD instructions. I'd use one of the 4 byte RGB formats.

**Comment by rr-** [comment]
Roger

**Comment by rr-** [comment]
I got it to work, but I get massive lags even with BGR32...

    static void draw_image(struct vo *vo, mp_image_t *mpi)                     
    {                                                                          
        struct priv *p = vo->priv;

        struct modeset_buf *front_buf = &p->dev->bufs[p->dev->front_buf];
        struct mp_image img = {0};
        mp_image_set_params(&img, &p->sws->dst);
        img.planes[0] = front_buf->map; //mmap()ed back buffer
        img.stride[0] = front_buf->stride; //for 720p video this value is a little above 5000

        mp_image_crop_rc(mpi, p->src_rc);
        mp_sws_scale(p->sws, &img, mpi);

        osd_draw_on_image(vo->osd, p->osd, mpi ? mpi->pts : 0, 0, &img);
    }

Did I get the following things right?

1. I do need to scale the image in every `draw_image()` call, as in, there's no resize-at-decoder-level magic cure.
2. If I experience tearing when enabling frame drops, adding v-sync to page flipping is going to help.
3. V-synced page flipping is said to happen asynchronously at the kernel level. This means I can use more buffers.

**Comment by wm4** [comment]
Try to find out what is slow by separating scaling and writing to the framebuffer. If the framebuffer is some of that freaky uncached gpu ram, of course scaling to it will be even slower than memcpy. System memcpy might also be slow (there's actually a SIMD memcpy specially for use with GPU mapped memory for dxva somewhere in the mpv tree, maybe you could try that).

If scaling to system ram is slow (though I doubt that, it's not THAT slow), try setting different swscale flags which are faster.

You are correct, there's no way to magically scale at decoder level (at least nothing reasonable).

**Comment by rr-** [comment]
When playing much less demanding file (640x360 `.avi` hardsubs), it dropped every other frame until I told it to use `--sws-scaler=point`, in which case it made **big** difference and played the file normally.

I used plain `memcpy` from RAM buffer to GPU's backbuffer. Using `gpu_memcpy` from `video/decoder/gpu_memcpy_sse4.c` failed with "illegal instruction" so I guess my CPU doesn't understand that. The next thing I'm going to try is to use `memcpy_pic` which skips strides, but I don't think it's going to make big difference. If that fails too, I'm going for triple buffering and asynchronous vsync.

**Comment by wm4** [comment]
You should profile what is taking how long.

**Comment by rr-** [comment]
Attempts to use callgrind show that most of stuff happens in `libsws` and `libav` though I'm not sure if I use it right.

Anyway, for some reason on 720p screen, when playing 720p video, I get very good performance if I add `-vf=scale=320:180 --sws-scaler=point`.

I don't understand it. The only possible explanation I see is as follows:

- Decoder finishes its work at 720p.
- VF downscales 720p decoder output to 180p in YUV color space. Only ~50k color space conversions need to happen.
- This image is then passed to VO.
- VO upscales input with point filter to 720p in RGB color space. This happens very fast, since no conversion need to be made, and because the scale factor is a power of 2, it allows cool optimizations.

I'd say that the current performance is comparable with `vo=x11`.

**Comment by wm4** [comment]
Running the Linux ``perf`` utility might also be a help. It doesn't emulate the CPU, so your program can run at realtime.

Is libswscale still accessing the actual framebuffer here?

**Comment by rr-** [comment]
Nope, it's scaling into a simple `char *buffer` that gets `memcpy()`d later on:

https://github.com/rr-/mpv/blob/drm/video/out/vo_drm.c#L363
https://github.com/rr-/mpv/blob/drm/video/out/vo_drm.c#L375

Like I said using the GPU-optimized `memcpy()` didn't work for me - I got segfault with `illegal instruction`.

**Comment by wm4** [comment]
Maybe try to steal a gpu memcpy from somewhere else (which isn't sse4). There should be tons of them around. MPlayer had one too.

**Comment by rr-** [comment]
`perf` says the number one offender is `mp_draw_sub_bitmaps`, then lots of calls in `libswscale.so`.

**Comment by wm4** [comment]
If there are no subtitles and no OSD, mp_draw_sub_bitmaps should be doing nothing.

**Comment by rr-** [comment]
There are subtitles in the test movie and I sometimes seek it, so OSD gets drawn.

Please note that my hardware is crap, too. Without HW acceleration it doesn't perform too well. Perhaps that's the case? 

#### CPU - AMD C-50
Key | Value
--- | ---
Clock Rate	| 1000 MHz
Level 1 Cache	| 128 KB
Level 2 Cache	| 1024 KB
Number of Cores / Threads	| 2 / 2
Max. Power Consumption (TDP = Thermal Design Power)	| 9 Watt
Manufacturing Technology	| 40 nm
Die Size	| 75 mm2
Socket	| FT1 BGA 413-Ball
Features	| MMX(+), SSE(1,2,3,3S,4A), AMD-V
64 Bit	| 64 Bit support
Hardware Virtualization	| AMD-V
Announcement Date	| 11/09/2010

#### GPU - Radeon HD 6250
Key | Value
--- | ---
Pipelines	| 80 - unified
Core Speed	| 280 MHz
Shared Memory	| yes
DirectX	| DirectX 11, Shader 5.0
technology	| 40 nm
Features	| DirectX 11, UVD3
Announcement Date	| 16.11.2010

**Comment by wm4** [comment]
I really don't know, but yes this hw looks a bit on the slower side to me. For what's it worth, with --vo=x11 on a core2 duo 2.33GHz at 1680x1050 it plays in realtime, even with the OSC visible, with some framedrops if subtitles cover much of the video.

**Comment by rr-** [comment]
I'm researching ways to integrate DRM + hardware acceleration. Turns out EGL is perfect candidate for this because it nicely proxies talking to 3rd party drivers, and is already used in other renderers (wayland, x11egl etc.). I succeeded to create `vo_drm_egl` (separate from `vo_drm`, because it imposes much more dependencies than `libdrm`) that draws simple triangles. :+1: But I'm afraid it's useless, because we want to draw 2D pictures, not OpenGL triangles. So...

1. Can I talk to EGL's frame buffer directly? I didn't find anything in `mpv` sources nor on the Internet, unfortunately.
2. Should I be doing this at all? Perhaps I should be creating a `mmap()`ed texture... or something?
3. Additionally, should I use `libswscale`, or should the EGL take care of the scaling? My guess is that this job belongs to EGL, because scaling and color space translations are exactly the kind of things that HW acceleration should help with.

Thanks for your guidance.

**Comment by wm4** [comment]
>Can I talk to EGL's frame buffer directly? I didn't find anything in mpv sources nor on the Internet, unfortunately.

I doubt it. It's probably for use with a OpenGL context, not anything else. But you could make a backend for vo_opengl; then vo_opengl would do all the rendering, while your code merely creates an opengl context for rendering to the screen. I guess gl_rpi.c is a good example for something that does almost nothing but create a GL context on the screen using EGL and some platform-specific low level APIs.

>Should I be doing this at all? Perhaps I should be creating a mmap()ed texture... or something?

Uh no, vo_opengl already has all the code to render video via textures. (Including yuv conversion etc.). But as I said, as a backend it would make sense.

**Comment by rr-** [comment]
I managed to draw first frame with proper size, aspect ratio and everything... then I get a segfault. This is because I haven't implemented buffer swapping yet. I'm afraid that making page flipping behave properly isn't going to be easy and will probably result in [something like this](https://github.com/GNOME/cogl/blob/master/cogl/winsys/cogl-winsys-egl-kms.c).

**Comment by wm4** [comment]
Uh, well, I can't really tell what all this cogl code does. Are you saying supporting this correctly requires loads of boilerplate?

**Comment by rr-** [comment]
Fortunately enough, page flipping in the cogl code constitutes for about 15% of the whole file size; the rest is initialization (which is already done) and integration with their modules. What I meant is that it's tricky to get it right, especially that DRM page flip events happen completely asynchronously (as opposed to simple double buffering).
I'll work on it in the upcoming week.

**Comment by wm4** [comment]
Thanks to rr- there's now a ``--vo=drm`` in git master.
